src/komet/kdist/soroban-semantics/data.md

```diff
@@ -344,5 +349,216 @@ module HOST-OBJECT
+
+```k
+    rule compare(A, B) => compareInt(ScValTypeOrd(A), ScValTypeOrd(B))
+      requires ScValTypeOrd(A) =/=Int ScValTypeOrd(B)
+```
+
+If `A` and `B` are of the same variant, they are compared by their underlying values.
+For scalar types the comparison is straightforward.
+
+```k
+    rule compare(SCBool(A), SCBool(B)) => compareBool(A, B)
+    rule compare(Void, Void)           => Equal
+    rule compare(Error(ATYP, ACODE), Error(BTYP, BCODE))
+      => #if ATYP ==K BTYP
+         #then compareInt(ACODE, BCODE)
+         #else compareInt(ErrorType2Int(ATYP), ErrorType2Int(BTYP))
+         #fi
+    rule compare(U32(A),  U32(B))  => compareInt(A, B)
+    rule compare(I32(A),  I32(B))  => compareInt(A, B)
+    rule compare(U64(A),  U64(B))  => compareInt(A, B)
+    rule compare(I64(A),  I64(B))  => compareInt(A, B)
+    rule compare(U128(A), U128(B)) => compareInt(A, B)
+    rule compare(I128(A), I128(B)) => compareInt(A, B)
+    rule compare(ScAddress(A), ScAddress(B)) => compareAddress(A, B)
+    rule compare(Symbol(A), Symbol(B))       => compareString(A, B)
+    rule compare(ScBytes(A), ScBytes(B))     => compareBytes(A, B)
+```
+
+For container types, the comparison is recursive as defined in `compareVec` and `compareMap`.
+
+- Maps are compared key-by-key in sorted or der of keys
+- Vectors are compared element-by-element in order
+
+```k
+    rule compare(ScVec(A), ScVec(B)) => compareVec(A, B)
+    rule compare(ScMap(A), ScMap(B)) => compareMap(A, sortedKeys(A), B, sortedKeys(B))
+```
+
+### Comparison of scalars
+
+```k
+    syntax Ordering ::= compareBool(Bool, Bool)             [function, total, symbol(compareBool)]
+ // ----------------------------------------------------------------------------------------------
+    rule compareBool(true, true)  => Equal
+    rule compareBool(true, false) => Greater
+    rule compareBool(false,true)  => Less
+    rule compareBool(false,false) => Equal
+
+    syntax Ordering ::= compareAddress(Address, Address)    [function, total, symbol(compareAddress)]
+ // -------------------------------------------------------------------------------------
+    rule compareAddress(Account(_),  Contract(_)) => Less
+    rule compareAddress(Contract(_), Account(_))  => Greater
+    rule compareAddress(Contract(A), Contract(B)) => compareBytes(A, B)
+    rule compareAddress(Account(A),  Account(B))  => compareBytes(A, B)
+
+    syntax Ordering ::= compareBytes(Bytes, Bytes)       [function, total, symbol(compareBytes)]
+                      | compareString(String, String)    [function, total, symbol(compareString)]
+ // ---------------------------------------------------------------------------------------------
+    rule compareBytes(A, B) => compareString(Bytes2String(A), Bytes2String(B))
+    rule compareString(A, B) => Less    requires A  <String B
+    rule compareString(A, B) => Equal   requires A ==String B
+    rule compareString(A, B) => Greater requires A  >String B
```

Comment:
It looks like the tests you added don't test string/bytes comparisons. Have you also verified that our semantics for String comparisons matches the comparisons for strings/bytes in soroban?

Reply:
@gtrepta I added a komet test for Bytes comparison after your review. I also implemented a new host function, `bytes_copy_to_linear_memory`, to support this test.